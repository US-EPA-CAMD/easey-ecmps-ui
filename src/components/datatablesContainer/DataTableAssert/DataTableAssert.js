import React, { useEffect, useState } from "react";

import Modal from "../../Modal/Modal";
import ModalDetails from "../../ModalDetails/ModalDetails";
import { DataTableRender } from "../../DataTableRender/DataTableRender";
import { extractUserInput } from "../../../additional-functions/extract-user-input";
import { modalViewData } from "../../../additional-functions/create-modal-input-controls";
import * as assertSelector from "../../../utils/selectors/assert";

import { Preloader } from "../../Preloader/Preloader";
import { connect } from "react-redux";
import { loadDropdowns } from "../../../store/actions/dropdowns";
import { convertSectionToStoreName } from "../../../additional-functions/data-table-section-and-store-names";

import {
  getActiveData,
  getInactiveData,
} from "../../../additional-functions/filter-data";

import {
  attachChangeEventListeners,
  removeChangeEventListeners,
  unsavedDataMessage,
} from "../../../additional-functions/prompt-to-save-unsaved-changes";

export const DataTableAssert = ({
  mdmData,
  loadDropdownsData,
  locationSelectValue,
  user,
  checkout,
  inactive,
  settingInactiveCheckBox,
  revertedState,
  setRevertedState,
  nonEditable,

  pagination,
  filter,
  controlInputs,
  controlDatePickerInputs,
  radioNames,
  payload,
  urlParameters,
  columnNames,
  dropdownArray,
  dataTableName,
  selectedLocation,
  showModal = false,
}) => {
  const [dataPulled, setDataPulled] = useState([]);
  const [show, setShow] = useState(showModal);
  const [selectedRow, setSelectedRow] = useState(null);
  const [selectedModalData, setSelectedModalData] = useState(null);
  const [dataLoaded, setDataLoaded] = useState(false);
  const [dropdownsLoaded, setDropdownsLoaded] = useState(false);

  const [updateTable, setUpdateTable] = useState(false);
  // need to test this part to fully test the page
  useEffect(() => {
    if (
      updateTable ||
      dataPulled.length <= 0 ||
      locationSelectValue ||
      revertedState
    ) {
      setDataLoaded(false);
      getDataTableApi(dataTableName, locationSelectValue, selectedLocation);

      setUpdateTable(false);
      setRevertedState(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [locationSelectValue, updateTable, revertedState, dataTableName]);

  // load dropdowns data (called once)
  useEffect(() => {
    if (
      mdmData.length === 0 &&
      mdmData.dropdowns[convertSectionToStoreName(dataTableName)]
    ) {
      loadDropdownsData(
        convertSectionToStoreName(dataTableName),
        dropdownArray
      );
      setDropdownsLoaded(true);
    } else {
      setDropdownsLoaded(true);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // get API for data
  // in  a timer because WAFS get takes a lil bit exxtra time to process, fixes update of datatable after editing data
  const getDataTableApi = (name, location, selectedLocationParameter) => {
    let timerFunc = setTimeout(() => {
      assertSelector
        .getDataTableApis(name, location, selectedLocationParameter)
        .then((res) => {
          setDataPulled(res.data);
          setDataLoaded(true);
        });
    }, [500]);
    setUpdateTable(true);
    return () => clearTimeout(timerFunc);
  };

  const saveData = () => {
    const userInput = extractUserInput(
      payload,
      ".modalUserInput",
      radioNames ? radioNames : null
    );

    assertSelector
      .saveDataSwitch(
        userInput,
        dataTableName,
        locationSelectValue,
        urlParameters
      )
      .then(() => {
        setShow(false);
        setUpdateTable(true);
      });
  };

  const createData = () => {
    const userInput = extractUserInput(
      payload,
      ".modalUserInput",
      radioNames ? radioNames : null
    );

    assertSelector
      .createDataSwitch(
        userInput,
        dataTableName,
        locationSelectValue,
        urlParameters
      )
      .then(() => {
        setShow(false);
        setUpdateTable(true);
      });
  };

  const [createNewData, setCreateNewData] = useState(false);

  // Executed when "View" action is clicked
  const openModal = (row, bool, create) => {
    let selectedData = null;
    setCreateNewData(create);
    if (dataPulled.length > 0 && !create) {
      selectedData = dataPulled.filter(
        (element) => element.id === row[`col${Object.keys(row).length - 1}`]
      )[0];
      setSelectedRow(selectedData);
    }
    setSelectedModalData(
      modalViewData(
        selectedData,
        controlInputs,
        controlDatePickerInputs,
        create,
        mdmData[convertSectionToStoreName(dataTableName)]
      )
    );
    setShow(true);

    setTimeout(() => {
      attachChangeEventListeners(".modalUserInput");
    });
  };

  const [viewBtn, setViewBtn] = useState(null);
  const [addBtn, setAddBtn] = useState(null);

  const closeModalHandler = () => {
    if (window.isDataChanged === true) {
      if (window.confirm(unsavedDataMessage) === true) {
        setShow(false);
        removeChangeEventListeners(".modalUserInput");
      }
    } else {
      setShow(false);
      removeChangeEventListeners(".modalUserInput");
    }
    if (addBtn) {
      addBtn.focus();
    }
  };

  const [dataSet, setDataSet] = useState([]);
  useEffect(() => {
    if (dataPulled.length > 0) {
      const activeOnly = getActiveData(dataPulled);
      const inactiveOnly = getInactiveData(dataPulled);

      // only active data >  disable checkbox and unchecks it
      if (activeOnly.length === dataPulled.length) {
        // uncheck it and disable checkbox
        //function parameters ( check flag, disable flag )
        settingInactiveCheckBox(false, true);
        setDataSet(
          assertSelector.getDataTableRecords(dataPulled, dataTableName)
        );
      }

      // only inactive data > disables checkbox and checks it
      if (inactiveOnly.length === dataPulled.length) {
        //check it and disable checkbox
        settingInactiveCheckBox(true, true);
        setDataSet(
          assertSelector.getDataTableRecords(dataPulled, dataTableName)
        );
      }
      // resets checkbox
      settingInactiveCheckBox(inactive[0], false);
      setDataSet(
        assertSelector.getDataTableRecords(
          !inactive[0] ? getActiveData(dataPulled) : dataPulled,
          dataTableName
        )
      );
    } else {
      setDataSet([]);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    dataPulled,
    inactive,
    locationSelectValue,
    dataTableName,
    updateTable,
    revertedState,
  ]);

  return (
    <div className="methodTable">
      <div className={`usa-overlay ${show ? "is-visible" : ""}`} />

      <DataTableRender
        openHandler={openModal}
        columnNames={columnNames}
        data={dataSet}
        dataLoaded={dataLoaded}
        pagination={pagination}
        filter={filter}
        actionsBtn={"View"}
        checkout={checkout}
        user={user}
        nonEditable={nonEditable}
        addBtn={openModal}
        addBtnName={`Create ${dataTableName}`}
        setViewBtn={setViewBtn}
        viewBtn={viewBtn}
        setAddBtn={setAddBtn}
        show={show}
      />
      {show ? (
        <Modal
          show={show}
          close={closeModalHandler}
          save={createNewData ? createData : saveData}
          showCancel={!(user && checkout) || nonEditable}
          showSave={user && checkout && !nonEditable}
          nonEditable={nonEditable}
          title={createNewData ? `Create ${dataTableName}` : `${dataTableName}`}
          exitBTN={createNewData ? `Create ${dataTableName}` : `Save and Close`}
          children={
            dropdownsLoaded ? (
              <div>
                <ModalDetails
                  modalData={selectedRow}
                  data={selectedModalData}
                  cols={2}
                  title={`${dataTableName}`}
                  viewOnly={!(user && checkout) || nonEditable}
                  create={createNewData}
                />
              </div>
            ) : (
              <Preloader />
            )
          }
        />
      ) : null}
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    mdmData: state.dropdowns,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    loadDropdownsData: (section, dropdownArray) => {
      dispatch(
        loadDropdowns(convertSectionToStoreName(section), dropdownArray)
      );
    },
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(DataTableAssert);
export { mapDispatchToProps };
export { mapStateToProps };
