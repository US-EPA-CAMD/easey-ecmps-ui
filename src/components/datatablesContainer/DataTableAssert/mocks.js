import { getMockMonitoringSpans } from "../../../mocks/functions";

export const getDataTableAssertProps = () => {
  return {
    mdmData: {
      testCode: [],
    },
    loadDropdownsData: jest.fn(),

    locationSelectValue: "6", //before
    user: "testuser",
    checkout: false,

    inactive: false,
    settingInactiveCheckBox: jest.fn(), //before
    revertedState: false,
    setRevertedState: jest.fn(), //before
    pagination: false,
    filter: false,
    // tabs comes from redux initial state
    tabs: [ // before
      {
        openedFacilityTabs: [],
        inactive: [true],
      },
    ],
    currentTabIndex: 0, // before

    controlInputs: {
      mecValue: ["MEC Value", "input", "", ""],
      mpcValue: ["MPC Value", "input", "", ""],
      mpfValue: ["MPF Value", "input", "", ""],
      spanValue: ["Span Value", "input", "", ""],
      fullScaleRange: ["Full Scale Range", "input", "", ""],
      scaleTransitionPoint: ["Scale Transition Point", "input", "", ""],
      defaultHighRange: ["Default High Range", "input", "", ""],
      flowSpanValue: ["Flow Span Value", "input", "", ""],
      flowFullScaleRange: ["Flow Full Scale Range", "input", "", ""],
      skip: ["", "skip", "", ""],
    },
    controlDatePickerInputs: {
      beginDate: ["Start Date", "date", "", ""],
      beginHour: ["Start Time", "hourDropdown", "", ""],
      endDate: ["End Date", "date", "", ""],
      endHour: ["End Time", "hourDropdown", "", ""],
    },
    radioName: null,
    payload: {
      locationId: 7,
      id: null,
      componentTypeCode: "string",
      spanScaleCode: "string",
      spanMethodCode: "string",
      mecValue: 0,
      mpcValue: 0,
      mpfValue: 0,
      spanValue: 0,
      fullScaleRange: 0,
      spanUnitsOfMeasureCode: "string",
      scaleTransitionPoint: "string",
      defaultHighRange: 0,
      flowSpanValue: 0,
      flowFullScaleRange: 0,
      beginDate: "2021-09-16T20:55:48.806Z",
      beginHour: 0,
      endDate: "2021-09-16T20:55:48.806Z",
      endHour: 0,
    },
    urlParameters: null,

    columnNames: ["Begin Date/Time", "End Date/Time"],
    dropdownArray: [[]],
    dataTableName: "Span",
    selectedLocation: 5,

    showModal: false,

    setUpdateRelatedTables: jest.fn(), //before

    updateRelatedTables: false,
  };
}

export const getMockBothActiveAndInactiveRecords = () => {
  const records = getMockMonitoringSpans()
  records[0].active = false
  return records
}

export const getMockOnlyInactiveRecords = () => {
  const records = getMockMonitoringSpans();
  const onlyInactive = records.map(span => {
    span.active = false
    return span
  })
  return onlyInactive
}

export const getMockMonitoringPlansUnitControlRecords = () => {
  const records = [
    {
      controlCode: "SNCR",
      originalCode: "0",
      installDate: "2005-01-01",
      optimizationDate: "2006-06-01",
      seasonalControlsIndicator: "0",
      retireDate: null,
      id: "1",
      unitId: 734,
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2009-02-20",
      active: true,
      parameterCode: "string"
    },
    {
      controlCode: "SNCR",
      originalCode: "0",
      installDate: "2005-01-01",
      optimizationDate: "2006-06-01",
      seasonalControlsIndicator: "0",
      retireDate: null,
      id: "2",
      unitId: 734,
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2009-02-20",
      active: true,
      parameterCode: "string"
    }
  ]
  return [...records]
}

export const getMockMonitoringPlansFuelDataRecords = () => {
  const records = [
    {
      fuelCode: "C",
      indicatorCode: "P",
      ozoneSeasonIndicator: 0,
      demGCV: "GGC",
      demSO2: "SGC",
      beginDate: "1995-01-01",
      endDate: "2015-03-30",
      id: "EXPCH0876-C97319A3E96540BF92077C58DB9E2A17",
      unitId: 91233,
      actualOrProjectCode: "",
      sulfurContent: "",
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2015-04-16",
      active: true
    },
    {
      fuelCode: "C",
      indicatorCode: "P",
      ozoneSeasonIndicator: 0,
      demGCV: "GGC",
      demSO2: "SGC",
      beginDate: "1995-01-01",
      endDate: "2015-03-30",
      id: "2",
      unitId: 91233,
      actualOrProjectCode: "",
      sulfurContent: "",
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2015-04-16",
      active: true
    }
  ]
  return [...records]
}

export const getMockUnitCapacity = () => {
  const records = [
    {
      maximumHourlyHeatInputCapacity: 2322.1,
      beginDate: "2009-01-01",
      endDate: null,
      id: "1",
      unitId: 4705,
      commercialOperationDate: "1977-04-01",
      date: "2019-01-01",
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2009-02-20",
      active: true,
      boilerTurbineType: "string"
    },
    {
      maximumHourlyHeatInputCapacity: 2322.1,
      beginDate: "2009-01-01",
      endDate: null,
      id: "2",
      unitId: 4705,
      commercialOperationDate: "1977-04-01",
      date: "2019-01-01",
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2009-02-20",
      active: true,
      boilerTurbineType: "string"
    },
  ]
  return [...records]
}

export const getMockLocationAttributes = () => {
  const records = [
    {
      ductIndicator: 1,
      bypassIndicator: 1,
      groundElevation: 40,
      stackHeight: 400,
      materialCode: "OTHER",
      shapeCode: "ROUND",
      crossAreaFlow: 306,
      crossAreaStackExit: 306,
      beginDate: "2008-01-01",
      endDate: null,
      id: "MELISSAMAT-E496678DE19E4827ADE4D541B18B5756",
      locationId: "5770",
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2009-02-20",
      active: true
    },
    {
      ductIndicator: 1,
      bypassIndicator: 1,
      groundElevation: 40,
      stackHeight: 400,
      materialCode: "OTHER",
      shapeCode: "ROUND",
      crossAreaFlow: 306,
      crossAreaStackExit: 306,
      beginDate: "2008-01-01",
      endDate: null,
      id: "MELISSAMAT-E496678DE19E4827ADE4D541B18B5756",
      locationId: "5770",
      userId: "abcde",
      addDate: "2009-02-20",
      updateDate: "2009-02-20",
      active: true
    },
  ]
  return [...records]
}
