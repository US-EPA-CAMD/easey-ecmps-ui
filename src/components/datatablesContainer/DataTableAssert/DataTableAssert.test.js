import React from "react";
import { shallow } from "enzyme";
import { render, waitForElement, fireEvent } from "@testing-library/react";
import { DataTableAssert } from "./DataTableAssert";
import { act } from "react-dom/test-utils";

jest.mock("axios");

const props = {
  locationSelectValue: "6",
  user: "testuser",
  checkout: false,
  inactive: false,
  settingInactiveCheckBox: () => {},
  revertedState: false,
  setRevertedState: () => {},
  pagination: false,
  filter: false,
  controlInputs: null,
  controlDatePickerInputs: null,
  radioName: "",
  payload: {},
  urlParameters: null,
  columnNames: ["", "", ""],
  dropdownArray: ["", "", ""],
  dataTableName: "",
  selectedLocation: "6",
  showModal: false,
};

describe("DataTableAssert", () => {
  it("should render and match snapshot", () => {
    const wrapper = shallow(<DataTableAssert {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render DataTableAssert", () => {
    act(() => {
      render(<DataTableAssert {...props} />);
    });
  });

  it("should fire test button", async () => {
    let { container } = await waitForElement(() =>
      render(<DataTableAssert {...props} />)
    );
    fireEvent.click(container.querySelector("#testBtn"));
  });
});
