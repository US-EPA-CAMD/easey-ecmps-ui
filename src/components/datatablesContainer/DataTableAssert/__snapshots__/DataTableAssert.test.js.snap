// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableAssert should render and match snapshot 1`] = `ShallowWrapper {}`;
